### PR TITLE
[Merged by Bors] - Support multiple BLS implementations

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

- Refactor the `bls` crate to support multiple BLS "backends" (e.g., milagro, blst, etc).
- Removes some duplicate, unused code in `common/rest_types/src/validator.rs`.
- Removes the old "upgrade legacy keypairs" functionality (these were unencrypted keys that haven't been supported for a few testnets, no one should be using them anymore).

## Additional Info

Most of the files changed are just inconsequential changes to function names.

## TODO

- [x] Optimization levels
- [x] Infinity point: https://github.com/supranational/blst/issues/11
- [x] Ensure milagro *and* blst are tested via CI
- [x] What to do with unsafe code?
- [x] Test infinity point in signature sets

### CODE REVIEW

File: crypto/bls/src/impls/blst.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+    fn public_key(&self) -> blst_core::PublicKey {
+        self.sk_to_pk()
+    }
+
+    fn sign(&self, msg: Hash256) -> blst_core::Signature {
+        self.sign(msg.as_bytes(), DST, &[])
+    }
+
+    fn serialize(&self) -> SecretHash {
```
Comment (adaszko) [inline crypto/bls/src/impls/blst.rs:266]: The name `SecretHash` could be more descriptive.  Something like `AutoZeroizingSecretKey` perhaps?

File: crypto/bls/src/generic_public_key_bytes.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    impl_serde_deserialize!();
+}
+
+impl<Pub> fmt::Debug for GenericPublicKeyBytes<Pub> {
+    impl_debug!();
+}
+
+#[cfg(feature = "arbitrary")]
+impl<Pub: 'static> arbitrary::Arbitrary for GenericPublicKeyBytes<Pub> {
+    impl_arbitrary!(PUBLIC_KEY_BYTES_LEN);
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply crypto/bls/src/generic_public_key_bytes.rs:149]: Good question. I've just re-implemented what was already existing.

@kirk-baird are you happy with this as-is?

File: crypto/bls/src/impls/blst.rs (hunk 1 above)
Comment (paulhauner) [reply crypto/bls/src/impls/blst.rs]: Good call! I named it `ZeroizeHash` :)

**Comment by paulhauner** [comment]
@adaszko thanks for the review, all comments addressed. Regarding the `Arbitrary`, if we don't hear from Kirk I think we just merge anyway since it's maintaining existing functionality.

File: crypto/bls/src/generic_public_key_bytes.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kirk-baird) [reply crypto/bls/src/generic_public_key_bytes.rs:149]: Yea that's a good point! However we are using arbitrary to do fuzzing, and when fuzzing it'll almost certainly have signature verification off (cause the chance of generating valid signatures is practically 0).

So I'm happy to leave as is.

**Comment by paulhauner** [comment]
I'm going to double-check this builds and works on my RasPi4 before merging...

**Comment by paulhauner** [comment]
> I'm going to double-check this builds and works on my RasPi4 before merging...

Yep, syncs Altona on my PI.

bors r+

**Pull request closed without merging**